draft-irtf-cfrg-aegis-aead.md

```diff
@@ -925,13 +938,536 @@ t = S3 ^ (LE64(ad_len_bits) || LE64(msg_len_bits))
+
+The state of a parallel mode is represented as a vector of AEGIS-128L or AEGIS-256 states.
+
+## Additional conventions and definitions
+
+- `D`: the parallelism degree
+- `R`: the absorption and output rate of the mode. With AEGIS-128X, the rate is `2 * 128 * D` bits. WIth AEGIS-256X, the rate is `128 * D` bits.
+- `V[j,i]`: the `j`-th AES block of the `i`-th state. `i` is in the `[0..D)` range. For AEGIS-128X, `j` is in the `[0..8)` range, while for AEGIS-256, `j` is in the `[0..6)` range.
+- `V'[j,i]`: the `j`-th AES block of the next `i`-th state.
+- `ctx`: the context separator.
+- `Byte(x)`: the value `x`, encoded as 8 bits.
+
+## Authenticated Encryption
+
+~~~
+Encrypt(msg, ad, key, nonce)
+~~~
+
+The `Encrypt` function of the parallel modes is similar to the AEGIS-128L and AEGIS-256 `Encrypt` function, but processes `R` bit input blocks per update.
+
+Steps:
+
+~~~
+Init(key, nonce)
+
+ct = {}
+
+ad_blocks = Split(ZeroPad(ad, R), R)
+for ai in ad_blocks:
+    Absorb(ai)
+
+msg_blocks = Split(ZeroPad(msg, R), R)
+for xi in msg_blocks:
+    ct = ct || Enc(xi)
+
+tag = Finalize(|ad|, |msg|)
+ct = Truncate(ct, |msg|)
+
+return ct and tag
+~~~
+
+## Authenticated Decryption
+
+~~~
+Decrypt(ct, tag, ad, key, nonce)
+~~~
+
+The `Decrypt` function of the parallel modes is similar to the AEGIS-128L and AEGIS-256 `Decrypt` function, but processes `R` bits per update.
+
+Steps:
+
+~~~
+Init(key, nonce)
+
+msg = {}
+
+ad_blocks = Split(ZeroPad(ad, R), R)
+for ai in ad_blocks:
+    Absorb(ai)
+
+ct_blocks = Split(ct, R)
+cn = Tail(ct, |ct| mod R)
+
+for ci in ct_blocks:
+    msg = msg || Dec(ci)
+
+if cn is not empty:
+    msg = msg || DecPartial(cn)
+
+expected_tag = Finalize(|ad|, |msg|)
+
+if CtEq(tag, expected_tag) is False:
+    erase msg
+    return "verification failed" error
+else:
+    return msg
+~~~
+
+## AEGIS-128X
+
+### The Init Function
+
+~~~
+Init(key, nonce)
+~~~
+
+The `Init` function initizalizes a vector of `D` AEGIS-128L states with the same `key` and `nonce`, but a different context `ctx`. The context is added to the state before every update.
+
+Steps:
+
+~~~
+for i in 0..D:
+    V[0,i] = key ^ nonce
+    V[1,i] = C1
+    V[2,i] = C0
+    V[3,i] = C1
+    V[4,i] = key ^ nonce
+    V[5,i] = key ^ C0
+    V[6,i] = key ^ C1
+    V[7,i] = key ^ C0
+
+nonce_v = {}
+key_v = {}
+for i in 0..D:
+    nonce_v = nonce_v || nonce
+    key_v = key_v || key
+
+Repeat(10,
+    for i in 0..D:
+        ctx = Byte(i)
+        V[3,i] = V[3,i] ^ ZeroPad(ctx)
+        V[7,i] = V[7,i] ^ ZeroPad(ctx)
+
+    Update(nonce_v, key_v)
+)
+~~~
+
+### The Update Function
+
+~~~
+Update(M0, M1)
+~~~
+
+The AEGIS-128X `Update` function is similar to the AEGIS-128L `Update` function, but absorbs `R` (`= 2 * 128 * D`) bits at once. `M0` and `M1` are `128 * D` bits instead of 128 bits. `M0` and `M1` are split into 128-bit blocks, each of them updating a different AEGIS-128L state.
```

Comment:
> > `R` (`= 2 * 128 * D`) bits
> 
> Not sure about this. How about moving the brackets after:
> 
> > `R` bits (`2 * 128 * D`)
> 
> Better or worse?

Moving the brackets after causes a unit mismatch. `2 * 128 * D` is a scalar, while `R bits` is a number of bits.

Reply:
> > `M0` and `M1` are `128 * D` bits instead of 128 bits. `M0` and `M1` are split into 128-bit blocks, each of them updating a different AEGIS-128L state.
> 
> Perhaps worth writing this without repeating `M0` and `M1`.

Indeed.

Reply:
> Moving the brackets after causes a unit mismatch

Ok let's leave it. Do you prefer it with or without the `=` would be my other question?

Reply:
The `=` doesn't seem necessary. Let's remove it.